error logging undefined method `backtrace_locations'

I was in the middle of troubleshooting an issue and created this file:

```ruby
require 'openhab'

    on_time = BackyardPlants_Watering_Tap.previous_state(:influxdb, skip_equal: true)
    logger.error(on_time.to_s)
```

The error logged was:
```
06:33:43.376 [ERROR] [ript.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/openhab/conf/automation/jsr223/ruby/personal/test.rb': org.jruby.embed.EvalFailedException: java.lang.IllegalArgumentException: Cannot format given Object as a Date
```

However when I put those two lines inside a rule:
```ruby
rule 'x' do
  on_start
  run do
    on_time = BackyardPlants_Watering_Tap.previous_state(:influxdb, skip_equal: true)
    logger.error(on_time.to_s)
  end
end
```

The error being logged was:
```
06:36:30.999 [ERROR] [ript.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/openhab/conf/automation/jsr223/ruby/personal/test.rb': org.jruby.embed.EvalFailedException: (NoMethodError) undefined method `backtrace_locations' for #<Java::JavaLang::IllegalArgumentException:0x42c7c72e>
In rule: x
<script>:9:in `<main>'
```

It seems that java exceptions don't have `backtrace_locations`? I also noticed that the exception message is lost here.

--- a/features/rule_language.feature
+++ b/features/rule_language.feature
@@ -108,8 +108,10 @@ Feature: rule_language
     When I deploy the rules file
     Then It should log 'OpenHAB ready for rule processing' within 5 seconds
 
+  @log_level_changed
   Scenario: Errors in a run block is logged with stack trace and exits rule
-    Given code in a rules file
+    Given log level INFO
+    And code in a rules file
       """
       def test
         test2
@@ -133,4 +135,25 @@ Feature: rule_language
     And It should log "<script>:8:in `test'" within 5 seconds
     And It should log "<script>:17:in `block in <main>'" within 5 seconds
     And It should log "<script>:15:in `<main>'" within 5 seconds
-    And It should not log 'This one works!' within 10 seconds
+    But It should not log 'This one works!' within 10 seconds
+
+  @log_level_changed
+  Scenario: Native java exceptions are handled
+    Given log level INFO
+    And code in a rules file
+      """
+      def test
+        Java::JavaLang::Integer.parseInt('k')
+      end
+
+      rule 'test' do
+        on_start
+        run { test }
+      end
+      """
+    When I deploy the rules file
+    Then It should log 'For input string: "k" (Java::JavaLang::NumberFormatException)' within 5 seconds
+    And It should log 'In rule: test' within 5 seconds
+    And It should log "RUBY.test(<script>:8)" within 5 seconds
+    And It should log "RUBY.<main>(<script>:13)" within 5 seconds
+    And It should log "RUBY.<main>(<script>:11)" within 5 seconds

--- a/features/step_definitions/openhab.rb
+++ b/features/step_definitions/openhab.rb
@@ -168,3 +168,13 @@ Given('metadata added to {string} in namespace {string}:') do |item, namespace, 
   response = Rest.add_metadata(item: item, namespace: namespace, config: config)
   raise "Response #{response.pretty_inspect} Request #{response.request.pretty_inspect}" unless response.success?
 end
+
+Given('(set )log level (to ){word}') do |level|
+  set_log_level('jsr223.jruby', level)
+  set_log_level('org.openhab.binding.jrubyscripting', level)
+  set_log_level('org.openhab.core.automation', level)
+end
+
+Given('log level for {word} is set to {word}') do |bundle, level|
+  set_log_level(bundle, level)
+end

--- a/features/support/env.rb
+++ b/features/support/env.rb
@@ -21,4 +21,10 @@ After('@reset_library') do
   start_openhab
 end
 
+After('@log_level_changed') do
+  set_log_level('jsr223.jruby', 'TRACE')
+  set_log_level('org.openhab.binding.jrubyscripting', 'TRACE')
+  set_log_level('org.openhab.core.automation', 'TRACE')
+end
+
 prepare_openhab

--- a/features/support/openhab.rb
+++ b/features/support/openhab.rb
@@ -98,3 +98,7 @@ def delete_items
   end
   wait_until(seconds: 30, msg: 'Items not empty') { Rest.items.length.zero? }
 end
+
+def set_log_level(bundle, level)
+  openhab_client("log:set #{level.upcase} #{bundle}")
+end

--- a/lib/openhab/log/logger.rb
+++ b/lib/openhab/log/logger.rb
@@ -22,8 +22,20 @@ module OpenHAB
       #
       # Regex for matching internal calls in a stack trace
       #
-      INTERNAL_CALL_REGEX = %r{(openhab-scripting-.*/lib)|(org/jruby/)}.freeze
+      INTERNAL_CALL_REGEX = %r{(openhab-scripting-.*/lib)|org[./]jruby}.freeze
       private_constant :INTERNAL_CALL_REGEX
+
+      #
+      # Regex for matching internal calls in a java stack trace
+      #
+      EXCLUDED_JAVA_PACKAGES = /jdk\.internal\.reflect|java\.lang\.reflect|org\.openhab|java\.lang\.Thread\.run/.freeze
+      private_constant :EXCLUDED_JAVA_PACKAGES
+
+      #
+      # Regex for matching internal calls in a java stack trace
+      #
+      JAVA_INTERNAL_CALL_REGEX = Regexp.union(INTERNAL_CALL_REGEX, EXCLUDED_JAVA_PACKAGES).freeze
+      private_constant :JAVA_INTERNAL_CALL_REGEX
 
       #
       # Create a new logger
@@ -60,8 +72,13 @@ module OpenHAB
       def clean_backtrace(error)
         return error if debug_enabled?
 
-        backtrace = error.backtrace_locations.reject { |line| INTERNAL_CALL_REGEX.match? line.to_s }
-        error.set_backtrace(backtrace.map(&:to_s))
+        if error.respond_to? :backtrace_locations
+          backtrace = error.backtrace_locations.map(&:to_s).reject { |line| INTERNAL_CALL_REGEX.match? line }
+          error.set_backtrace(backtrace)
+        elsif error.respond_to? :stack_trace
+          backtrace = error.stack_trace.reject { |line| JAVA_INTERNAL_CALL_REGEX.match? line.to_s }
+          error.set_stack_trace(backtrace)
+        end
         error
       end
 

